COO intersection: allow external hash + hash reuse in sparse_mask

nikitaved opened the pull request:
External hash implies more flexibility in the op coverage + performance improvement.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95165
* #94991
* __->__ #94596
* #94406

nikitaved committed:
COO intersection: allow external hash + hash reuse in sparse_mask

[ghstack-poisoned]

nikitaved committed:
Update on "COO intersection: allow external hash + hash reuse in sparse_mask"




[ghstack-poisoned]

nikitaved committed:
Update on "COO intersection: allow external hash + hash reuse in sparse_mask"

External hash implies more flexibility in the op coverage.
Next up: `index_select`.




[ghstack-poisoned]

nikitaved commented on aten/src/ATen/native/sparse/SparseBinaryOpIntersectionCommon.h:
```diff
@@ -249,13 +267,22 @@ void _sparse_binary_op_intersection_kernel_impl(
 
-    auto hash = at::empty({probably_coalesced._nnz()},
-        indices.options().dtype(kLong));
+    auto hash = source_indices_hash_opt.has_value()
+      ? (*source_indices_hash_opt).contiguous()
+      : at::empty({probably_coalesced._nnz()}, indices.options().dtype(kLong));
+    const auto* RESTRICT hash_ptr = source_indices_hash_opt.has_value()
+      ? hash.data_ptr<int64_t>()
+      : nullptr;
```
Bug: wrong place to put it.

nikitaved committed:
Update on "COO intersection: allow external hash + hash reuse in sparse_mask"

External hash implies more flexibility in the op coverage.
Next up: `index_select`.




[ghstack-poisoned]

nikitaved committed:
Update on "COO intersection: allow external hash + hash reuse in sparse_mask"

External hash implies more flexibility in the op coverage.
Next up: `index_select`.




[ghstack-poisoned]

nikitaved committed:
Update on "COO intersection: allow external hash + hash reuse in sparse_mask"

External hash implies more flexibility in the op coverage.
Next up: `index_select`.




[ghstack-poisoned]

nikitaved committed:
Update on "COO intersection: allow external hash + hash reuse in sparse_mask"

External hash implies more flexibility in the op coverage.
Next up: `index_select`.




[ghstack-poisoned]

nikitaved committed:
Update on "COO intersection: allow external hash + hash reuse in sparse_mask"

External hash implies more flexibility in the op coverage.
Next up: `index_select`.




[ghstack-poisoned]

nikitaved committed:
Update on "COO intersection: allow external hash + hash reuse in sparse_mask"

External hash implies more flexibility in the op coverage.
Next up: `index_select`.




[ghstack-poisoned]

nikitaved committed:
Update on "COO intersection: allow external hash + hash reuse in sparse_mask"

External hash implies more flexibility in the op coverage.
Next up: `index_select`.




[ghstack-poisoned]

nikitaved committed:
Update on "COO intersection: allow external hash + hash reuse in sparse_mask"

External hash implies more flexibility in the op coverage.
Next up: `index_select`.




[ghstack-poisoned]

nikitaved committed:
Update on "COO intersection: allow external hash + hash reuse in sparse_mask"

External hash implies more flexibility in the op coverage.
Next up: `index_select`.




[ghstack-poisoned]

nikitaved committed:
Update on "COO intersection: allow external hash + hash reuse in sparse_mask"

External hash implies more flexibility in the op coverage.
Next up: `index_select`.




[ghstack-poisoned]

nikitaved committed:
Update on "COO intersection: allow external hash + hash reuse in sparse_mask"

External hash implies more flexibility in the op coverage.
Next up: `index_select`.




[ghstack-poisoned]

nikitaved committed:
Update on "COO intersection: allow external hash + hash reuse in sparse_mask"

External hash implies more flexibility in the op coverage.
Next up: `index_select`.




[ghstack-poisoned]

nikitaved committed:
Update on "COO intersection: allow external hash + hash reuse in sparse_mask"

External hash implies more flexibility in the op coverage.
Next up: `index_select`.




[ghstack-poisoned]

nikitaved committed:
Update on "COO intersection: allow external hash + hash reuse in sparse_mask"

External hash implies more flexibility in the op coverage.
Next up: `index_select`.




[ghstack-poisoned]

nikitaved committed:
Update on "COO intersection: allow external hash + hash reuse in sparse_mask"

External hash implies more flexibility in the op coverage.
Next up: `index_select`.




[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/sparse/SparseStubs.h:
```diff
@@ -2,17 +2,20 @@
 
 namespace native {
 
+using OptTensor = c10::optional<Tensor>;
```
Nit: Adding this to the header makes it too widely available.

nikitaved commented:
@cpuhrsch , yes, it allows to implement intersections along a specific dimension, and so is broadcasting all with the same kernel.

cpuhrsch reviewed (approved):
Can we remove `using OptTensor` from the headers? Otherwise this is good to go.

nikitaved committed:
Update on "COO intersection: allow external hash + hash reuse in sparse_mask"

External hash implies more flexibility in the op coverage + performance improvement.




[ghstack-poisoned]

nikitaved commented:
@pytorchbot merge

nikitaved commented on aten/src/ATen/native/sparse/SparseBinaryOpIntersectionCommon.h:
```diff
@@ -146,10 +148,23 @@ void _sparse_binary_op_intersection_kernel_impl(
+    // No need to coalesce in such a case.
+    if (distributive_with_sum) {
+      return std::make_tuple(t, t_hash_opt);
+    } else {
+      // Otherwise coalesce and force hash recompute.
+      return std::make_tuple(t.coalesce(), static_cast<OptTensor>(c10::nullopt));
+    }
```
TODO: no need to recompute, just sufficient to wrap into a sparse tensor and coalesce.

nikitaved commented:
@pytorchbot merge -m "unrelated failure"

nikitaved commented:
@pytorchbot merge -f "unrelated failure"
